code/addons/docs/src/manifest.ts

```diff
@@ -0,0 +1,185 @@
+import { logger } from 'storybook/internal/node-logger';
+import type {
+  ComponentManifest,
+  DocsIndexEntry,
+  IndexEntry,
+  Manifests,
+  Path,
+  PresetPropertyFn,
+  StorybookConfigRaw,
+} from 'storybook/internal/types';
+
+const ATTACHED_MDX_TAG = 'attached-mdx';
+const UNATTACHED_MDX_TAG = 'unattached-mdx';
+
+export interface DocsManifestEntry {
+  id: string;
+  name: string;
+  path: Path;
+  title: string;
+  content?: string;
+  error?: { name: string; message: string };
+}
+
+interface DocsManifest {
+  v: number;
+  docs: Record<string, DocsManifestEntry>;
+}
+
+interface ComponentManifestWithDocs extends ComponentManifest {
+  docs?: Record<string, DocsManifestEntry>;
+}
+
+interface ComponentsManifestWithDocs {
+  v: number;
+  components: Record<string, ComponentManifestWithDocs>;
+}
+
+interface ManifestsWithDocs extends Manifests {
+  docs?: DocsManifest;
+  components?: ComponentsManifestWithDocs;
+}
+
+/** Converts a DocsIndexEntry to a DocsManifestEntry by reading its file content. */
+export async function createDocsManifestEntry(entry: DocsIndexEntry): Promise<DocsManifestEntry> {
+  const absolutePath = path.join(process.cwd(), entry.importPath);
+  try {
+    const content = await fs.readFile(absolutePath, 'utf-8');
+    return {
+      id: entry.id,
+      name: entry.name,
+      path: entry.importPath,
+      title: entry.title,
+      content,
+    };
+  } catch (err) {
+    return {
+      id: entry.id,
+      name: entry.name,
+      path: entry.importPath,
+      title: entry.title,
+      error: {
+        name: err instanceof Error ? err.name : 'Error',
+        message: err instanceof Error ? err.message : String(err),
+      },
+    };
+  }
+}
+
+/**
+ * Extracts unattached MDX entries (standalone docs not attached to any component). These are added
+ * to a separate `docs` manifest.
+ */
+export async function extractUnattachedDocsEntries(
+  entries: DocsIndexEntry[]
+): Promise<Record<string, DocsManifestEntry>> {
+  if (entries.length === 0) {
+    return {};
+  }
+  const entriesWithContent = await Promise.all(entries.map(createDocsManifestEntry));
+  return Object.fromEntries(entriesWithContent.map((entry) => [entry.id, entry]));
+}
+
+/**
+ * Extracts attached docs entries by adding them to their corresponding component manifests.
+ *
+ * Returns the updated components manifest with docs added to each component.
+ */
+export async function extractAttachedDocsEntries(
+  entries: DocsIndexEntry[],
+  existingComponents: ComponentsManifestWithDocs | undefined
+): Promise<ComponentsManifestWithDocs | undefined> {
+  if (!existingComponents || entries.length === 0) {
+    return existingComponents;
+  }
+
+  const entriesWithContent = await Promise.all(entries.map(createDocsManifestEntry));
+
+  // Add docs to their corresponding components based on the entry id prefix
+  for (const docsEntry of entriesWithContent) {
+    const componentId = docsEntry.id.split('--')[0];
```

Comment:
We should use a helper function to get the component id. I think the helper function `toId` from `storybook/internal/csf` could be helpful here.